### DISCUSSION TITLE
Autocompletion and docstring support in text editors and IDEs

### DISCUSSION THREAD

**Discussion by wigging** [wjakob/nanobind#224] [General]
I use Sublime Text and Visual Studio Code to write Python code. When I import a package or module into a Python file, these editors will provide autocompletion for functions, classes, and other objects that are available from the package or module. The editors will also show the docstring associated with that function, class, or object. This [page](https://code.visualstudio.com/docs/python/editing) for Visual Studio Code demonstrates autocompletion and docstring support in the editor. When I import a module created by nanobind, I don't get autocompletion or docstring support for that module in the text editor or IDE. I can view the docstrings in a REPL like iPython or an interactive Python session but they don't show up in editors. Does anyone else have this problem or is this not possible for nanobind modules?

**Comment by wjakob** [comment]
You will need to generate stubs. Take a look at this: https://github.com/wjakob/nanobind/discussions/163

**Reply by wigging** [reply]
I read that discussion and it looks like the stub file (`.pyi`) generators are mainly for providing type information via Python type hints. But what about the docstrings? If I use a stub file generator it looks like I still have to manually add the docstrings to the generated `.pyi` files.

I use the [NumPy style](https://numpydoc.readthedocs.io/en/v1.5.0/example.html) for docstrings in my Python code. This style has sections for parameters, examples, returns, etc. I don't see how you would do this for nanobind or pybind11 bindings without using `.pyi` files that contain NumPy styled docstrings. Is there a way to write this type of docstring in the C++ binding code and then have the stub file generator create the `.pyi` files with the docstrings?

**Reply by dylan-baros** [reply]
@wigging ever figure this out?

**Reply by wigging** [reply]
Nope.

**Reply by wjakob** [reply]
Nanobind will copy the docstrings (based on whatever you specified in the C++ bindings) into the stub. Tools like VSCode will pick them up from there.
